File: app/src/main/java/com/budgetbuddy/ui/budget/BudgetViewModel.kt
package com.budgetbuddy.ui.budget

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.budgetbuddy.data.local.entities.BudgetEntity
import com.budgetbuddy.data.local.entities.CategoryEntity
import com.budgetbuddy.data.local.entities.TransactionType
import com.budgetbuddy.data.repository.BudgetRepository
import com.budgetbuddy.data.repository.CategoryRepository
import com.budgetbuddy.data.repository.TransactionRepository
import com.budgetbuddy.util.DateUtils
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

data class BudgetWithSpend(
    val budget: BudgetEntity,
    val category: CategoryEntity,
    val spent: Double,
    val progressPercent: Int,
    val status: BudgetStatus
)

enum class BudgetStatus { OK, WARNING, EXCEEDED }

@HiltViewModel
class BudgetViewModel @Inject constructor(
    private val budgetRepository: BudgetRepository,
    private val categoryRepository: CategoryRepository,
    private val transactionRepository: TransactionRepository
) : ViewModel() {

    private val _budgetsWithSpend = MutableStateFlow<List<BudgetWithSpend>>(emptyList())
    val budgetsWithSpend: StateFlow<List<BudgetWithSpend>> = _budgetsWithSpend.asStateFlow()

    private val _saveState = MutableStateFlow<Boolean?>(null)
    val saveState: StateFlow<Boolean?> = _saveState.asStateFlow()

    private var budgetJob: kotlinx.coroutines.Job? = null

    fun loadBudgets(userId: String) {
        if (budgetJob?.isActive == true) return
        budgetJob = viewModelScope.launch {
            val month = DateUtils.currentMonth()
            val year  = DateUtils.currentYear()
            val start = DateUtils.startOfMonth()
            val end   = DateUtils.endOfMonth()

            combine(
                budgetRepository.getBudgetsForMonth(userId, month, year),
                transactionRepository.getTransactionsByDateRange(userId, start, end)
            ) { budgets, transactions ->
                val spendByCat = transactions
                    .filter { it.type == TransactionType.EXPENSE }
                    .groupBy { it.categoryId }
                    .mapValues { (_, txs) -> txs.sumOf { it.amount } }
                budgets.mapNotNull { budget ->
                    val category = categoryRepository.getCategoryById(budget.categoryId) ?: return@mapNotNull null
                    val spent = spendByCat[budget.categoryId] ?: 0.0
                    val pct = if (budget.limitAmount > 0) ((spent / budget.limitAmount) * 100).toInt().coerceIn(0, 150) else 0
                    val status = when {
                        pct >= 100 -> BudgetStatus.EXCEEDED
                        pct >= 80  -> BudgetStatus.WARNING
                        else       -> BudgetStatus.OK
                    }
                    BudgetWithSpend(budget, category, spent, pct, status)
                }
            }.collect { _budgetsWithSpend.value = it }
        }
    }

    fun saveBudget(userId: String, categoryId: Long, limitAmount: Double) {
        viewModelScope.launch {
            val budget = BudgetEntity(
                userId = userId,
                categoryId = categoryId,
                limitAmount = limitAmount,
                month = DateUtils.currentMonth(),
                year  = DateUtils.currentYear()
            )
            budgetRepository.insertOrUpdateBudget(budget)
            _saveState.value = true
        }
    }

    fun deleteBudget(budget: BudgetEntity) {
        viewModelScope.launch { budgetRepository.deleteBudget(budget) }
    }
}

File: ios/BudgetBuddy/ViewModels/BudgetViewModel.swift
import Foundation

enum BudgetStatus { case ok, warning, exceeded, underMin }

struct BudgetWithSpend: Identifiable {
    var id: UUID { budget.id }
    let budget: BBBudget
    let category: BBCategory
    let spent: Double
    let progressPercent: Int
    let status: BudgetStatus
}

@MainActor
final class BudgetViewModel: ObservableObject {
    @Published var budgetsWithSpend: [BudgetWithSpend] = []
    @Published var categories: [BBCategory]            = []
    @Published var isLoading  = false
    @Published var saveResult: Bool?
    @Published var error: String?

    private let budgetRepo:      BudgetRepository
    private let categoryRepo:    CategoryRepository
    private let transactionRepo: TransactionRepository

    init(budgetRepo: BudgetRepository, categoryRepo: CategoryRepository,
         transactionRepo: TransactionRepository) {
        self.budgetRepo      = budgetRepo
        self.categoryRepo    = categoryRepo
        self.transactionRepo = transactionRepo
    }

    func load(userId: String) {
        isLoading = true
        Task {
            do {
                let cal   = Calendar.current
                let month = cal.component(.month, from: .now)
                let year  = cal.component(.year,  from: .now)
                let start = DateUtils.startOfMonth()
                let end   = DateUtils.endOfMonth()

                let budgets    = try budgetRepo.forMonth(userId: userId, month: month, year: year)
                let allCats    = try categoryRepo.all()
                let catMap     = Dictionary(uniqueKeysWithValues: allCats.map { ($0.id, $0) })
                categories     = allCats

                let transactions = try transactionRepo.forDateRange(userId: userId, start: start, end: end)
                let spendByCat   = Dictionary(grouping: transactions.filter { $0.type == .expense }, by: { $0.categoryId })
                    .mapValues { $0.reduce(0) { $0 + $1.amount } }

                budgetsWithSpend = budgets.compactMap { budget in
                    guard let cat = catMap[budget.categoryId] else { return nil }
                    let spent = spendByCat[budget.categoryId] ?? 0
                    let pct   = budget.limitAmount > 0
                        ? min(150, Int((spent / budget.limitAmount) * 100)) : 0
                    let status: BudgetStatus = pct >= 100 ? .exceeded
                                             : pct >= 80  ? .warning
                                             : (budget.minAmount > 0 && spent < budget.minAmount) ? .underMin
                                             : .ok
                    return BudgetWithSpend(budget: budget, category: cat, spent: spent,
                                           progressPercent: pct, status: status)
                }
            } catch { self.error = error.localizedDescription }
            isLoading = false
        }
    }

    func save(userId: String, categoryId: UUID, limitAmount: Double,
              minAmount: Double = 0, completion: (() -> Void)? = nil) {
        Task {
            do {
                let cal   = Calendar.current
                try await budgetRepo.insertOrUpdate(userId: userId, categoryId: categoryId,
                                                    limitAmount: limitAmount, minAmount: minAmount,
                                                    month: cal.component(.month, from: .now),
                                                    year:  cal.component(.year,  from: .now))
                saveResult = true
                load(userId: userId)
                completion?()
            } catch {
                saveResult  = false
                self.error  = error.localizedDescription
            }
        }
    }

    func delete(_ budget: BBBudget, userId: String) {
        Task {
            try? await budgetRepo.delete(budget)
            load(userId: userId)
        }
    }
}
